adapters/repos/db/index_queue.go

```diff
@@ -423,31 +463,103 @@ func (q *IndexQueue) indexer() {
 	}
 }
 
-func (q *IndexQueue) pushToWorkers(max int, wait bool) {
-	chunks := q.queue.borrowChunks(max)
-	for i, c := range chunks {
-		select {
-		case <-q.ctx.Done():
-			// release unsent borrowed chunks
-			for _, c := range chunks[i:] {
-				q.queue.releaseChunk(c)
+func (q *IndexQueue) pushToWorkers(max int, wait bool) int {
+	chunks := q.queue.dequeue(max)
+	if len(chunks) == 0 {
+		return 0
+	}
+
+	var minID uint64
+
+	doneChans := make([]chan struct{}, 0, len(chunks))
+
+	for i := range chunks {
+		c := chunks[i]
+		ids := make([]uint64, 0, c.cursor)
+		vectors := make([][]float32, 0, c.cursor)
+		var deleted []uint64
+
+		for i := range c.data[:c.cursor] {
+			if minID == 0 || c.data[i].id < minID {
+				minID = c.data[i].id
 			}
 
-			return
+			if q.queue.IsDeleted(c.data[i].id) {
+				deleted = append(deleted, c.data[i].id)
+			} else {
+				ids = append(ids, c.data[i].id)
+				vectors = append(vectors, c.data[i].vector)
+			}
+		}
+
+		if len(ids) == 0 {
+			q.Logger.Debug("all vectors in the chunk are deleted. skipping")
+			continue
+		}
+
+		done := make(chan struct{})
+		doneChans = append(doneChans, done)
+
+		select {
+		case <-q.ctx.Done():
+			return i
 		case q.indexCh <- job{
-			chunk:   c,
 			indexer: q.Index,
-			queue:   q.queue,
 			ctx:     q.ctx,
+			ids:     ids,
+			vectors: vectors,
+			done:    done,
 		}:
+			q.queue.ResetDeleted(deleted...)
 		}
 	}
 
+	q.doneChans.Lock()
+	q.doneChans.chans = append(q.doneChans.chans, doneChans...)
+	q.doneChans.Unlock()
+
+	q.queue.persistCheckpoint(minID)
+
 	if wait {
-		q.queue.wait(q.ctx)
+		q.wait(q.ctx)
+	}
+
+	return len(chunks)
+}
+
+// waits for all workers to finish their current tasks.
+func (q *IndexQueue) wait(ctx context.Context) {
+	q.doneChans.Lock()
+	chans := q.doneChans.chans
+	q.doneChans.Unlock()
```

Comment:
Can we use RLock here?

Reply:
It's not really necessary tbh, as only two goroutines can have access to the chans:
- whoever calls `Close`
- the indexer goroutine
In that configuration Mutex is enough and is much faster than RWMutex, although in our case lock speed is not important